libmamba/src/core/subdirdata.cpp

```diff
@@ -160,17 +160,53 @@ namespace mamba
 
     }
 
+    subdirdata_error::subdirdata_error(const char* msg)
+        : m_message(msg)
+    {
+    }
+
+    subdirdata_error::subdirdata_error(const std::string& msg)
+        : m_message(msg)
+    {
+    }
+
+    const std::string& subdirdata_error::what() const noexcept
+    {
+        return m_message;
+    }
+
+    auto MSubdirData::create(const Channel& channel,
+                             const std::string& platform,
+                             const std::string& url,
+                             MultiPackageCache& caches,
+                             const std::string& repodata_fn) -> expected<MSubdirData>
+    {
+        try
+        {
+            return MSubdirData(channel, platform, url, caches, repodata_fn);
+        }
+        catch(std::exception& e)
+        {
+            return tl::make_unexpected(e.what());
+        }
+        catch(...)
+        {
+            return tl::make_unexpected("Unkown error when trying to load subdir data " + url);
+        }
+    }
+
     MSubdirData::MSubdirData(const Channel& channel,
                              const std::string& platform,
                              const std::string& url,
                              MultiPackageCache& caches,
```

Comment:
should we make this a shared ptr?

Reply:
I don't think there is a need for a `shared_ptr` here, the subdirdata is not a shared ressource (meaning we know when to delete it and who should do it).

Reply:
I was thinking we could make the `MultiPackageCache` a `shared_ptr` so we know that the pointer doesn't become invalid.